### PR TITLE
Enable .NET 8 for uwp MultiTarget, upgrade packages

### PR DESCRIPTION
**Pull request by Arlodotexe**
For components that declare a `uwp` MultiTarget, this PR:
- Introduces the `net8.0-windows10.0.26100.0` TFM alongside the existing `uap10.0.17763` 
- Auto-enables the required properties for targeting modern dotnet in uwp
- Updates several platform packages to versions that support modern dotnet on uwp
- Adds several workarounds and fixes to get builds working again with all the latest tooling

### CODE REVIEW

File: MultiTarget/MultiTargetIdentifiers.props [hunk 1]
```diff
@@ -57,7 +57,10 @@
       This property is only for changing the version used by Uno.
       Force the version to 2 for UWP and 3 for WinAppSDK.
      -->
-    <WinUIMajorVersion Condition="'$(HasWinUI)' == 'true' AND '$(IsUwp)' == 'true'">2</WinUIMajorVersion>
-    <WinUIMajorVersion Condition="'$(HasWinUI)' == 'true' AND '$(IsWinAppSdk)' == 'true'">3</WinUIMajorVersion>
+    <WinUIMajorVersion Condition="'$(HasWinUI)' == 'true' AND '$(IsUwp)' == 'true' AND $(WinUIMajorVersion) == ''">2</WinUIMajorVersion>
+    <WinUIMajorVersion Condition="'$(HasWinUI)' == 'true' AND '$(IsWinAppSdk)' == 'true' AND $(WinUIMajorVersion) == ''">3</WinUIMajorVersion>
+
+    <UseUwp Condition="'$(HasWinUI)' == 'true' AND '$(IsUwp)' == 'true' AND $([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">true</UseUwp>
+    <UseUwpTools Condition="'$(HasWinUI)' == 'true' AND '$(IsUwp)' == 'true' AND $([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">true</UseUwpTools>
```
Comment (Sergio0694) [inline MultiTarget/MultiTargetIdentifiers.props:64]: This isn't needed, VS 17.12 will automatically set this one whenever `UseUwp` is set

File: MultiTarget/NoWarn.props [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<Project>  
+  <!-- Disable warnings for 'This call site is reachable on all platforms.' caused by underlying platform. -->
+  <PropertyGroup Condition="'$(IsUwp)' == 'true' AND $([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">
+    <NoWarn>$(NoWarn);CA1416;</NoWarn>
```
Comment (Sergio0694) [inline MultiTarget/NoWarn.props:4]: Want to suppress `NETSDK1219` too? It warns that UWP .NET 9 is in preview.

File: MultiTarget/PackageReferences/Uwp.props [hunk 3]
```diff
@@ -1,5 +1,6 @@
 <Project>
   <ItemGroup>
-    <PackageReference Condition="'$(HasWinUI)' == 'true'" Include="Microsoft.UI.Xaml" Version="2.7.0" />
+    <PackageReference Condition="'$(HasWinUI)' == 'true'" Include="Microsoft.UI.Xaml" Version="2.8.7-prerelease.241119001" />
+    <PackageReference Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'" Include="Microsoft.Windows.CsWinRT" Version="2.1.3" PrivateAssets="all" />
```
Comment (Sergio0694) [inline MultiTarget/PackageReferences/Uwp.props:4]: Bump this to 2.1.6

File: MultiTarget/WinUI.Extra.props [hunk 4]
```diff
@@ -27,6 +27,13 @@
     <None PackagePath="lib/net7.0-windows10.0.18362" Include="$(MSBuildThisFileDirectory)/_._" Pack="true" />
   </ItemGroup>
 
+  <PropertyGroup Condition="'$(IsUwp)' == 'true'">
+    <Platforms>x86;x64;arm64</Platforms>
+    
+    <WindowsSdkPackageVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">10.0.26100.39</WindowsSdkPackageVersion>
```
Comment (Sergio0694) [inline MultiTarget/WinUI.Extra.props:33]: Use .54, not .39

**Review by Sergio0694 — CHANGES_REQUESTED** [review]
Looks good! Just left some notes 🙂

File: MultiTarget/PackageReferences/Uwp.props (hunk 3 above)
Comment (Arlodotexe) [reply MultiTarget/PackageReferences/Uwp.props:4]: Just here or for all references to `Microsoft.Windows.CsWinRT`?

File: MultiTarget/WinUI.Extra.props (hunk 4 above)
Comment (Arlodotexe) [reply MultiTarget/WinUI.Extra.props:33]:
```suggestion
    <WindowsSdkPackageVersion Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">10.0.26100.54</WindowsSdkPackageVersion>
```

File: MultiTarget/PackageReferences/Uwp.props (hunk 3 above)
Comment (Sergio0694) [reply MultiTarget/PackageReferences/Uwp.props:4]: All the ones in Toolkit **libraries**. For samples (ie. when the CsWinRT ref is not transitive), you can also use 2.2.0.

File: MultiTarget/MultiTargetIdentifiers.props (hunk 1 above)
Comment (Arlodotexe) [reply MultiTarget/MultiTargetIdentifiers.props:64]:
```suggestion
```

File: MultiTarget/NoWarn.props (hunk 2 above)
Comment (Arlodotexe) [reply MultiTarget/NoWarn.props:4]:
```suggestion
    <NoWarn>$(NoWarn);CA1416;NETSDK1219;</NoWarn>
```

File: MultiTarget/PackageReferences/Uwp.props (hunk 3 above)
Comment (Arlodotexe) [reply MultiTarget/PackageReferences/Uwp.props:4]:
```suggestion
    <PackageReference Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'" Include="Microsoft.Windows.CsWinRT" Version="2.1.6" PrivateAssets="all" />
```

File: Build-Toolkit-Components.ps1 [hunk 5]
```diff
@@ -104,6 +104,20 @@ if ($null -eq $ExcludeMultiTargets)
...
+    if ($WinUIMajorVersion -eq 2)
+    {
+        $ExcludeMultiTargets = $ExcludeMultiTargets + 'wasdk'
+    }
+    else
+    {
+        $ExcludeMultiTargets = $ExcludeMultiTargets + 'uwp'
+    }
+}
+
```
Comment (michael-hawker) [inline Build-Toolkit-Components.ps1:120]: Seems like we're doing this in 3 places, should we have a helper function and start making another helper script these all include and reference? (maybe https://learn.microsoft.com/en-us/powershell/scripting/learn/ps101/10-script-modules ? )

File: ProjectHeads/GenerateSingleSampleHeads.ps1 [hunk 6]
```diff
@@ -1,30 +1,48 @@
...
-
     This script also generates a solution for the project and will open Visual Studio.
+
 .PARAMETER componentPath
     Folder for the project to copy the project heads to.
-.PARAMETER heads
-    Which heads to include to copy, defaults to all. (Currently ignored.)
+
+.PARAMETER MultiTargets
+    Specifies the MultiTarget TFM(s) to include for building the components. The default value is 'all'.
```
Comment (michael-hawker) [inline ProjectHeads/GenerateSingleSampleHeads.ps1:13]: This is not the default, the default is (uwp, wasm, and wasdk)

Comment (Arlodotexe) [reply ProjectHeads/GenerateSingleSampleHeads.ps1:13]:
```suggestion
    Specifies the MultiTarget TFM(s) to include for building the components. The default value is 'uwp', 'wasm', 'wasdk'.
```

**Comment by Arlodotexe** [comment]
It would appear that adding the modern .NET for UWP TFM results in our gallery app being unbuildable in uap mode:
```
4>C:\Users\arlog\.nuget\packages\microsoft.windows.sdk.cpp\10.0.26100.2454\c\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(383,13): error MSB4064: The "UsingCsWinRT" parameter is not supported by the "CompileXaml" task loaded from assembly: Microsoft.Windows.UI.Xaml.Build.Tasks, Version=0.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 from the path: C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Build.Tasks.dll. Verify that the parameter exists on the task, the <UsingTask> points to the correct assembly, and it is a settable public instance property.
4>C:\Users\arlog\.nuget\packages\microsoft.windows.sdk.cpp\10.0.26100.2454\c\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(354,5): error MSB4063: The "CompileXaml" task could not be initialized with its input parameters.
4>Done building project "Converters.Samples.csproj" -- FAILED.
```

Starting with the 10.0.20348 SDK, the version of the tasks assembly is always 0.0.0.0, which is why VS can't tell the different newer versions apart.

An issue has been filed for the platform [here](https://developercommunity.visualstudio.com/t/VS-caching-of-build-task-dll-s-breaks/10807077) (thanks to @Sergio0694 for helping here).

In the meantime, we'll have to employ a workaround to get the gallery buildable again. The code should be functionally identical to `uap10.*`, and NativeAoT changes should be functionally identical to those made for the Windows App SDK and Uno Platform.

Since this is just an issue with Visual Studio, the workaround we're looking at is to disable modern .NET for UWP for local development and only enable it for build in the CI. When the issue is resolved, we can enable it again for local testing and build.

**Comment by Arlodotexe** [comment]
Looks like [this](https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/235/commits/218a656261914f654c0d936677fc6617c15f166b) has fixed our problem with Sample and Source component projects not compiling in Visual Studio, but similar issues remain on the head itself. 

The package `Microsoft.Windows.SDK.CPP` is being used to bring us SDK changes early and may be part of the problem/solution here. 

A few observations:

As of 218a656261914f654c0d936677fc6617c15f166b, building the Sample Gallery for uwp results in:
```
3>C:\Users\arlog\.nuget\packages\microsoft.windows.sdk.cpp\10.0.26100.2454\c\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(354,5): Xaml Internal Error error WMC9999: Object reference not set to an instance of an object.
========== Build: 2 succeeded, 1 failed, 2 up-to-date, 0 skipped ==========
```

If we revert 7625ac54c837e19b33e97c4abd93cecf10d32306 and only include this package on modern .net (not uap), we get the error message:
```
3>C:\Users\arlog\.nuget\packages\microsoft.windows.sdk.cpp\10.0.26100.2454\c\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(383,13): error MSB4064: The "UsingCsWinRT" parameter is not supported by the "CompileXaml" task loaded from assembly: Microsoft.Windows.UI.Xaml.Build.Tasks, Version=0.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35 from the path: C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Build.Tasks.dll. Verify that the parameter exists on the task, the <UsingTask> points to the correct assembly, and it is a settable public instance property.
3>C:\Users\arlog\.nuget\packages\microsoft.windows.sdk.cpp\10.0.26100.2454\c\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(354,5): error MSB4063: The "CompileXaml" task could not be initialized with its input parameters. 
========== Build: 2 succeeded, 1 failed, 2 up-to-date, 0 skipped ==========
```

Removing this package reference entirely results in an error similar to the latest commit:
```
3>C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(322,5): Xaml Internal Error error WMC9999: Object reference not set to an instance of an object.
```

**Comment by Arlodotexe** [comment]
Seems like a proper fix will require a bugfix from the platform, we'll need to continue searching for a workaround here.

218a656261914f654c0d936677fc6617c15f166b was good progress and did fix some things, but we're finding that the changes here are still causing problems when building the UWP head.  Locally, when running a build in VS with modern .NET on UWP seemingly _completely removed_ (no TFMs, no `UseUwp`, no extra package references), we still get the build error:
```
51>(MarkupCompilePass1 target) -> 
51>  C:\Program Files (x86)\Windows Kits\10\bin\10.0.26100.0\XamlCompiler\Microsoft.Windows.UI.Xaml.Common.targets(322,5): Xaml Internal Error error WMC9999: Object reference not set to an instance of an object.
```

Main builds fine, so there's additional changes in this branch that are causing this. Investigating.

File: Build-Toolkit-Components.ps1 (hunk 5 above)
Comment (Arlodotexe) [reply Build-Toolkit-Components.ps1:120]: There are chunks of these files with overlapping functionality, not just this. For now, the code should be largely aligned to make it easier to extract later. We can file a ticket for it, if you like?

File: MultiTarget/MultiTargetIdentifiers.props [hunk 7]
```diff
@@ -43,7 +43,8 @@
 
     <IsUwp Condition="'$(IsUwp)' == '' AND $(UwpTargetFramework.Contains('$(TargetFramework);')) AND '$(UwpTargetFramework)' != '' AND '$(MultiTargetsUwp)' == 'true'">true</IsUwp>
     <IsWinAppSdk Condition="'$(IsWinAppSdk)' == '' AND $(WinAppSdkTargetFramework.Contains('$(TargetFramework);')) AND '$(WinAppSdkTargetFramework)' != '' AND '$(MultiTargetsWasdk)' == 'true'">true</IsWinAppSdk>
-    <IsWindowsTfm Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">true</IsWindowsTfm>
+    <TargetPlatformIdentifier>$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)'))</TargetPlatformIdentifier>
```
Comment (Sergio0694) [inline MultiTarget/MultiTargetIdentifiers.props:46]: Don't name it `TargetPlatformIdentifier`, better not to mess with well known properties set by the .NET SDK. Just name it something like `MultiTargetPlatformIdentifier` or whatever.

**Review by Sergio0694 — CHANGES_REQUESTED** [review]

File: MultiTarget/MultiTargetIdentifiers.props (hunk 7 above)
Comment (Arlodotexe) [reply MultiTarget/MultiTargetIdentifiers.props]: Good catch, fixed in https://github.com/CommunityToolkit/Tooling-Windows-Submodule/pull/235/commits/c0d535d20494566b135b8f527e98d0b7b918ecbc

**Comment by Arlodotexe** [comment]
CI is blocked by this error. Cause unknown, needs investigation.

image

File: MultiTarget/PackageReferences/Uwp/Microsoft.UI.Xaml.props [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<Project>
+	<!-- Condition must be on a <When> statement in its own file for this to work in Visual Studio. https://stackoverflow.com/a/16557059 -->
+	<Choose>
+    <!-- Temp workaround until next Windows SDK servicing update -->
```
Comment (Arlodotexe) [inline MultiTarget/PackageReferences/Uwp/Microsoft.UI.Xaml.props:4]:
```suggestion
```

File: MultiTarget/PackageReferences/Uwp/Microsoft.Windows.CsWinRT.props [hunk 9]
```diff
@@ -0,0 +1,13 @@
+<Project>
+	<!-- Condition must be on a <When> statement in its own file for this to work in Visual Studio. https://stackoverflow.com/a/16557059 -->
+	<Choose>
+    <!-- Temp workaround until next Windows SDK servicing update -->
```
Comment (Arlodotexe) [inline MultiTarget/PackageReferences/Uwp/Microsoft.Windows.CsWinRT.props:4]:
```suggestion
```

**Review by michael-hawker — APPROVED** [review]

**Review by Sergio0694 — APPROVED** [review]
:shipit:

**Pull request merged**